File: HW/HW11/HW11.py
def floodFill(self, image, sr, sc, color):
    def dfs(image, sr, sc, color, newColor):
        if sr < 0 or sr >= len(image) or sc < 0 or sc >= len(image[0]) or image[sr][sc] != color:
            return
        image[sr][sc] = newColor
        dfs(image, sr + 1, sc, color, newColor)
        dfs(image, sr - 1, sc, color, newColor)
        dfs(image, sr, sc + 1, color, newColor)
        dfs(image, sr, sc - 1, color, newColor)
    
    if image[sr][sc] != color:
        dfs(image, sr, sc, image[sr][sc], color)
    return image

# The Python solution utilizes a recursive Depth-First Search (DFS) approach to traverse the image pixels connected to the starting pixel.
# It defines a helper function dfs that recursively visits neighboring pixels, updating their color if they match the starting pixel's color.
# The process continues until all connected pixels of the same color are visited.
# If the starting pixel's color is different from the specified color, the function invokes the DFS to perform the flood fill.
# Finally, it returns the modified image after the flood fill operation.

File: HW/HW11/HW11.java
public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
    int color = image[sr][sc];
    if (color != newColor) {
        dfs(image, sr, sc, color, newColor);
    }
    return image;
}

private void dfs(int[][] image, int sr, int sc, int color, int newColor) {
    if (sr < 0 || sr >= image.length || sc < 0 || sc >= image[0].length || image[sr][sc] != color) {
        return;
    }
    image[sr][sc] = newColor;
    dfs(image, sr + 1, sc, color, newColor);
    dfs(image, sr - 1, sc, color, newColor);
    dfs(image, sr, sc + 1, color, newColor);
    dfs(image, sr, sc - 1, color, newColor);
}

// The solution uses a recursive approach Depth-First Search (DFS) to traverse the connected pixels starting from the given pixel.
// It defines a method called dfs that recursively visits neighboring pixels and updates their color if they match the starting pixel's color.
// The process continues until all connected pixels of the same color are visited.
// If the starting pixel's color is different from the specified color, the floodFill method initiates the DFS to perform the flood fill operation.
// The modified image is returned after the flood fill operation.
